File: app/src/main/kotlin/com/lifehub/app/api/ApiService.kt
package com.lifehub.app.api

import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

// ==================== WEATHER API ====================
data class WeatherResponse(
    val main: MainWeather,
    val weather: List<Weather>,
    val wind: Wind,
    val name: String
)

data class MainWeather(
    val temp: Double,
    val humidity: Int,
    val pressure: Int
)

data class Weather(
    val main: String,
    val description: String,
    val icon: String
)

data class Wind(
    val speed: Double
)

interface WeatherApi {
    @GET("weather")
    suspend fun getWeather(
        @Query("q") city: String,
        @Query("appid") apiKey: String = "YOUR_API_KEY",
        @Query("units") units: String = "metric"
    ): Response<WeatherResponse>
}

// ==================== NEWS API ====================
data class NewsResponse(
    val articles: List<Article>
)

data class Article(
    val title: String,
    val description: String?,
    val url: String,
    val urlToImage: String?,
    val publishedAt: String
)

interface NewsApi {
    @GET("v2/top-headlines")
    suspend fun getNews(
        @Query("country") country: String = "us",
        @Query("apiKey") apiKey: String = "YOUR_API_KEY"
    ): Response<NewsResponse>
}

// ==================== CURRENCY API ====================
data class CurrencyResponse(
    val rates: Map<String, Double>
)

interface CurrencyApi {
    @GET("v6/YOUR_API_KEY/latest/{base}")
    suspend fun getRates(
        @Path("base") base: String
    ): Response<CurrencyResponse>
}

// ==================== CRYPTO API ====================
data class CryptoResponse(
    val id: String,
    val name: String,
    val current_price: Double,
    val price_change_percentage_24h: Double,
    val image: String
)

interface CryptoApi {
    @GET("coins/markets")
    suspend fun getCrypto(
        @Query("vs_currency") currency: String = "usd",
        @Query("order") order: String = "market_cap_desc",
        @Query("per_page") perPage: Int = 20,
        @Query("page") page: Int = 1
    ): Response<List<CryptoResponse>>
}

// ==================== TRANSLATION API ====================
data class TranslationResponse(
    val translatedText: String
)

interface TranslationApi {
    @GET("translate")
    suspend fun translate(
        @Query("q") text: String,
        @Query("source") source: String,
        @Query("target") target: String
    ): Response<TranslationResponse>
}

// ==================== COUNTRIES API ====================
data class Country(
    val name: Name,
    val capital: List<String>?,
    val population: Long,
    val region: String,
    val flags: Flags,
    val currencies: Map<String, CurrencyInfo>?
)

data class Name(
    val common: String,
    val official: String
)

data class Flags(
    val png: String
)

data class CurrencyInfo(
    val name: String,
    val symbol: String?
)

interface CountriesApi {
    @GET("v3.1/all")
    suspend fun getAllCountries(): Response<List<Country>>
}

// ==================== NASA API ====================
data class NasaResponse(
    val title: String,
    val explanation: String,
    val url: String,
    val date: String,
    val media_type: String
)

interface NasaApi {
    @GET("planetary/apod")
    suspend fun getApod(
        @Query("api_key") apiKey: String = "DEMO_KEY"
    ): Response<NasaResponse>
}

// ==================== FUN APIs ====================
data class JokeResponse(
    val setup: String?,
    val delivery: String?,
    val joke: String?,
    val type: String
)

interface JokeApi {
    @GET("joke/Any")
    suspend fun getRandomJoke(): Response<JokeResponse>
}

data class FactResponse(
    val text: String
)

interface FactApi {
    @GET("random.json?language=en")
    suspend fun getRandomFact(): Response<FactResponse>
}

data class QuoteResponse(
    val content: String,
    val author: String
)

interface QuoteApi {
    @GET("random")
    suspend fun getRandomQuote(): Response<QuoteResponse>
}

File: lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';

class ApiService {
  static const String weatherBaseUrl = 'https://api.openweathermap.org/data/2.5';
  static const String newsBaseUrl = 'https://newsapi.org/v2';
  static const String cryptoBaseUrl = 'https://api.coingecko.com/api/v3';
  static const String countriesBaseUrl = 'https://restcountries.com/v3.1';
  static const String nasaBaseUrl = 'https://api.nasa.gov/planetary';
  static const String jokeBaseUrl = 'https://v2.jokeapi.dev/joke';
  static const String factBaseUrl = 'https://uselessfacts.jsph.pl/api/v2/facts';
  static const String quoteBaseUrl = 'https://api.quotable.io';
  static const String translateBaseUrl = 'https://libretranslate.de';
  static const String currencyBaseUrl = 'https://api.exchangerate-api.com/v4/latest';

  // API Keys - يمكنك تغييرها لاحقاً
  static const String weatherApiKey = 'YOUR_API_KEY';
  static const String newsApiKey = 'YOUR_API_KEY';
  static const String nasaApiKey = 'DEMO_KEY';

  // ========== WEATHER ==========
  static Future<Weather> getWeather(String city) async {
    final response = await http.get(
      Uri.parse('$weatherBaseUrl/weather?q=$city&appid=$weatherApiKey&units=metric'),
    );
    
    if (response.statusCode == 200) {
      return Weather.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load weather');
    }
  }

  // ========== NEWS ==========
  static Future<List<NewsArticle>> getNews() async {
    final response = await http.get(
      Uri.parse('$newsBaseUrl/top-headlines?country=us&apiKey=$newsApiKey'),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final List articles = data['articles'];
      return articles.map((e) => NewsArticle.fromJson(e)).toList();
    } else {
      throw Exception('Failed to load news');
    }
  }

  // ========== CRYPTO ==========
  static Future<List<Crypto>> getCrypto() async {
    final response = await http.get(
      Uri.parse('$cryptoBaseUrl/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=20&page=1'),
    );
    
    if (response.statusCode == 200) {
      final List data = jsonDecode(response.body);
      return data.map((e) => Crypto.fromJson(e)).toList();
    } else {
      throw Exception('Failed to load crypto');
    }
  }

  // ========== COUNTRIES ==========
  static Future<List<Country>> getCountries() async {
    final response = await http.get(
      Uri.parse('$countriesBaseUrl/all'),
    );
    
    if (response.statusCode == 200) {
      final List data = jsonDecode(response.body);
      return data.map((e) => Country.fromJson(e)).toList();
    } else {
      throw Exception('Failed to load countries');
    }
  }

  // ========== NASA ==========
  static Future<NasaApod> getNasaApod() async {
    final response = await http.get(
      Uri.parse('$nasaBaseUrl/apod?api_key=$nasaApiKey'),
    );
    
    if (response.statusCode == 200) {
      return NasaApod.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load NASA APOD');
    }
  }

  // ========== JOKE ==========
  static Future<Joke> getJoke() async {
    final response = await http.get(
      Uri.parse('$jokeBaseUrl/Any'),
    );
    
    if (response.statusCode == 200) {
      return Joke.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load joke');
    }
  }

  // ========== FACT ==========
  static Future<Fact> getFact() async {
    final response = await http.get(
      Uri.parse('$factBaseUrl/random?language=en'),
    );
    
    if (response.statusCode == 200) {
      return Fact.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load fact');
    }
  }

  // ========== QUOTE ==========
  static Future<Quote> getQuote() async {
    final response = await http.get(
      Uri.parse('$quoteBaseUrl/random'),
    );
    
    if (response.statusCode == 200) {
      return Quote.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load quote');
    }
  }

  // ========== TRANSLATE ==========
  static Future<String> translate(String text, String source, String target) async {
    final response = await http.post(
      Uri.parse('$translateBaseUrl/translate'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'q': text,
        'source': source,
        'target': target,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return data['translatedText'] ?? 'Translation error';
    } else {
      throw Exception('Failed to translate');
    }
  }

  // ========== CURRENCY ==========
  static Future<Map<String, dynamic>> getExchangeRates(String base) async {
    final response = await http.get(
      Uri.parse('$currencyBaseUrl/$base'),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return data['rates'];
    } else {
      throw Exception('Failed to load exchange rates');
    }
  }
}
